stdlib/join/merge_join.go

```diff
@@ -0,0 +1,522 @@
+// It makes the following assumptions about the input tables:
+//   * They have the same group key
+//   * They are sorted by the columns in the `on` parameter
+//   * We are performing an equijoin
+type MergeJoinTransformation struct {
+	ctx                     context.Context
+	on                      []ColumnPair
+	as                      *JoinFn
+	left, right             execute.DatasetID
+	leftSchema, rightSchema []flux.ColMeta
+	method                  string
+	d                       *execute.TransportDataset
+	mu                      sync.Mutex
+	mem                     memory.Allocator
+
+	leftFinished,
+	rightFinished bool
+}
+
+func NewMergeJoinTransformation(
+	id execute.DatasetID,
+	s plan.ProcedureSpec,
+	a execute.Administration,
+) (*MergeJoinTransformation, error) {
+	spec, ok := s.(*SortMergeJoinProcedureSpec)
+	if !ok {
+		return nil, errors.New(codes.Internal, "unsupported join spec - not a sortMergeJoin")
+	}
+	mem := a.Allocator()
+	return &MergeJoinTransformation{
+		ctx:    a.Context(),
+		on:     spec.On,
+		as:     NewJoinFn(spec.As),
+		left:   a.Parents()[0],
+		right:  a.Parents()[1],
+		method: spec.Method,
+		d:      execute.NewTransportDataset(id, mem),
+		mem:    mem,
+	}, nil
+}
+
+func (t *MergeJoinTransformation) ProcessMessage(m execute.Message) error {
+	defer m.Ack()
+
+	switch m := m.(type) {
+	case execute.ProcessChunkMsg:
+		chunk := m.TableChunk()
+		id := m.SrcDatasetID()
+		state, _ := t.d.Lookup(chunk.Key())
+		s, ok, err := t.processChunk(chunk, state, id)
+		if err != nil {
+			return err
+		}
+		if ok {
+			t.d.Set(chunk.Key(), s)
+		}
+	case execute.FlushKeyMsg:
+		id := m.SrcDatasetID()
+		key := m.Key()
+		state, _ := t.d.Lookup(key)
+		s, _ := state.(*joinState)
+		if id == t.left {
+			s.left.done = true
+		} else if id == t.right {
+			s.right.done = true
+		}
+
+		if s.finished() {
+			if err := t.flush(s); err != nil {
+				return err
+			}
+		}
+	case execute.FinishMsg:
+		err := m.Error()
+		if err != nil {
+			t.d.Finish(err)
+			return nil
+		}
+
+		id := m.SrcDatasetID()
+		if id == t.left {
+			t.leftFinished = true
+		} else if id == t.right {
+			t.rightFinished = true
+		}
+
+		if t.isFinished() {
+			err = t.d.Range(func(key flux.GroupKey, value interface{}) error {
+				s, ok := value.(*joinState)
+				if !ok {
+					return errors.New(codes.Internal, "received bad joinState")
+				}
+				s.left.schema = t.leftSchema
+				s.right.schema = t.rightSchema
+				return t.flush(s)
+			})
+			t.d.Finish(err)
+		}
+	default:
+		return errors.New(codes.Internal, "invalid message")
+	}
+	return nil
+}
+
+func (t *MergeJoinTransformation) initState(state interface{}) (*joinState, bool) {
+	if state != nil {
+		s, ok := state.(*joinState)
+		return s, ok
+	}
+	s := joinState{}
+	return &s, true
+}
+
+func (t *MergeJoinTransformation) processChunk(chunk table.Chunk, state interface{}, id execute.DatasetID) (*joinState, bool, error) {
```

Comment:
A high-level description of what happens when you process a chunk would be great here.